lib/CSV/HistoryPlayer.pm

```diff
@@ -0,0 +1,134 @@
+package CSV::HistoryPlayer;
+
+use List::MoreUtils qw(uniq);
+use Moo;
+use Path::Tiny;
+use Text::CSV;
+
+use strictures 2;
+use namespace::clean;
+
+our $VERSION = '0.01';
+
+has 'root_dir' => (
+    is       => 'ro',
+    required => 1
+);
+
+has 'dir_filter' => (
+    is      => 'ro',
+    default => sub {
+        return sub { 1 }
+    });
+
+has 'files_maper' => (
+    is      => 'ro',
+    default => sub {
+        return sub {
+            my $files = shift;
+            return [sort { $a cmp $b } @$files];
+            }
+    });
+
+has 'files' => (is => 'lazy');
+
+has _current_data => (is => 'rw');
+
+has '_reader' => (is => 'lazy');
+
+sub _build_files {
+    my $self = shift;
+
+    my @files;
+    my @dirs_queue = (path($self->root_dir));
+    my $dir_filter = $self->dir_filter;
+    while (@dirs_queue) {
+        my $dir = shift(@dirs_queue);
+        if ($dir_filter->($dir)) {
+            for my $c ($dir->children) {
+                push @dirs_queue, $c if (-d $c);
+                push @files, $c
+                    if ($c =~ /\.csv$/i && -s $c && -r $c && -f $c);
```

Comment:
I suspected that something like that exists, but I didn't know about `_` magic :)